src/celephem/spiceinterface.cpp

```diff
@@ -27,56 +33,152 @@ namespace
 
 using ResidentKernelsSet = std::set<fs::path>;
 
-ResidentKernelsSet* getResidentKernelsSet()
+#ifdef _WIN32
+using LibraryHandle = HMODULE;
+#else
+using LibraryHandle = void*;
+#endif
+
+struct SpiceLibrary
 {
-    // Track loaded SPICE kernels in order to avoid loading the same kernel
-    // multiple times. This is a static variable because SPICE uses a global
-    // kernel pool.
-    static ResidentKernelsSet* residentKernelsSet = new std::set<fs::path>;
-    return residentKernelsSet;
-}
+    explicit SpiceLibrary(LibraryHandle _handle) : handle(_handle) {}
+    ~SpiceLibrary();
 
-} // end unnamed namespace
+    SpiceInterface spiceInterface;
+    ResidentKernelsSet residentKernelsSet;
+    LibraryHandle handle;
+};
 
-/*! Perform one-time initialization of SPICE.
- */
+SpiceLibrary::~SpiceLibrary()
+{
+#ifdef _WIN32
+    FreeLibrary(handle);
+#else
+    dlclose(handle);
+#endif
+}
+
+template<typename T>
 bool
-InitializeSpice()
+SetInterfacePtr(LibraryHandle handle, T& destination, const char* name)
 {
-    // Set the error behavior to the RETURN action, so that
-    // Celestia do its own handling of SPICE errors.
-    erract_c("SET", 0, (SpiceChar*)"RETURN");
+#ifdef _WIN32
+    destination = reinterpret_cast<T>(GetProcAddress(handle, name));
+#else
+    destination = reinterpret_cast<T>(dlsym(handle, name));
+#endif
+    if (destination == nullptr)
+    {
+        GetLogger()->error("Could not find function {} in cspice library\n", name);
+        return false;
+    }
 
     return true;
 }
 
+std::unique_ptr<SpiceLibrary>
+LoadSpiceLibrary()
+{
+#if defined(_WIN32)
+    HMODULE handle = LoadLibraryExW(L"plugins\\cspice.dll", NULL, LOAD_LIBRARY_SEARCH_APPLICATION_DIR);
```

Comment:
On UWP we need to use [LoadPackagedLibrary](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-loadpackagedlibrary) for UWP. We can check for UWP [here](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/migrate-to-windows-app-sdk/misc-info).

Reply:
I've restructured the code to allow third-party front-ends to supply their own loading implementation.